aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -173,6 +173,37 @@ TORCH_META_FUNC(scatter_add)
   scatter_meta_impl(*this, self, dim, index, src, "add");
 }
 
+TORCH_PRECOMPUTE_META_FUNC(index_add)
+(const Tensor& self, int64_t dim, const Tensor& index, const Tensor& source, const Scalar& alpha) {
+  dim = maybe_wrap_dim(dim, self.dim());
+  auto numel = index.numel();
+
+  TORCH_CHECK_INDEX(index.dim() <= 1, "index_add_(): Index is supposed to be a vector, but got dim: ",
+                    index.dim(), " with type: ", index.scalar_type(), " and size: ", index.sizes());
+  TORCH_CHECK(index.scalar_type() == ScalarType::Long || index.scalar_type() == ScalarType::Int,
+              "index_add_(): Expected dtype int32/int64 for index but got: ", index.scalar_type());
+  TORCH_CHECK(self.scalar_type() == source.scalar_type(),
+              "index_add_(): self (", self.scalar_type(), ") and source (", source.scalar_type(),
+              ") must have the same scalar type");
+  TORCH_CHECK(dim == 0 || dim < source.dim(),
+              "index_add_(): Indexing dim ", dim, " is out of bounds of the source tensor with dim ",
+              source.dim());
+  TORCH_CHECK(numel == (source.dim() == 0 ? 1 : source.size(dim)),
+              "index_add_(): Number of indices (", numel, ") should be equal to source.size(dim): (",
+              source.size(dim), "), for dim: ", dim);
+
+  auto& result = maybe_get_output(0);
+  bool is_defined = result.defined();
+  set_output(self.sizes(), self.options());
+  if (is_defined) {
+    at::assert_no_internal_overlap(result);
+    at::assert_no_overlap(result, index);
+    at::assert_no_overlap(result, source);
+  }
+
+  return TORCH_PRECOMPUTE_STRUCT(index_add)().set_dim(dim);
```

Comment:
Just a question for me to learn. What's all this business with `TORCH_PRECOMPUTE_STRUCT` and `TORCH_PRECOMPUTE_META_FUNC`? I have not read abut this in the RFC. When should I use these?

Reply:
This was something which was added after the RFC release. To quote from the PR (https://github.com/pytorch/pytorch/pull/61746) that added this change:

> This commit introduces a new feature for structured kernels that allows
kernels to declare quantities as "precomputed" in
native_functions.yaml, compute them once in the meta function and
reuse them again in the impl.

There was a lot of redundancy, specially with usage of `maybe_wrap_dim` in both meta and impl functions, and I've mostly used this `TORCH_PRECOMPUTE_<STRUCT/META_FUNC>` to avoid redundancy (it, of course, can be used for other variables as well). These will also require some changes in the `native_functions.yaml` file.

If you don't have a quantity being computed in the meta function, and re-used in the impl function, then this is not required. I hope this helps answer your question.